### PR TITLE
Book a timeslot functionality

### LINKED ISSUE #12: Book a timeslot

### PR DESCRIPTION
**Pull request by KiraBurova**

### CODE REVIEW

File: user/handler.go [hunk 1]
```diff
@@ -54,24 +54,45 @@ func Register(w http.ResponseWriter, r *http.Request) {
...
-	// TODO: work on book time implementation
+	repo := NewRepository(db.DbInstance)
+
+	var data BookTimeData
+	json.NewDecoder(r.Body).Decode(&data)
+
+	user, err := repo.GetById(data.Id)
+
+	if err != nil {
+		w.WriteHeader(http.StatusBadRequest)
```
Comment (rutaka-n) [inline user/handler.go]: is it always a bad request?

File: user/repository.go [hunk 2]
```diff
@@ -68,5 +68,15 @@ func (r Repository) UserExists(username string) (bool, error) {
 	return false, err
 }
 
-func (r Repository) Update(id int, user User) {}
-func (r Repository) Delete(id int)            {}
+func (r Repository) Update(user User) error {
```
Comment (rutaka-n) [inline user/repository.go]: should we improve the db schema design to better handle book time logic?

**Review by rutaka-n — APPROVED** [review]
LGTM 👍, but I left a few points to discuss.

File: user/handler.go [hunk 3]
```diff
@@ -54,24 +54,45 @@ func Register(w http.ResponseWriter, r *http.Request) {
...
 
-	// params := mux.Vars(r)
-	// userId := params["userId"]
+	if slot, ok := timeSlots[data.Time]; ok {
+		var booked = slot.(map[string]interface{})["Booked"]
 
-	// user, err := repo.GetById(userId)
+		if booked == true {
+			// or 409 Conflict ?
+			w.WriteHeader(http.StatusForbidden)
```
Comment (rutaka-n) [inline user/handler.go]: better to choose a bad request or conflict.

File: user/handler.go [hunk 4]
```diff
@@ -54,24 +54,45 @@ func Register(w http.ResponseWriter, r *http.Request) {
 
 }
 
+type BookTimeData struct {
+	Id   string `json:"id"`
+	Time string `json:"time"`
+}
+
 func BookTime(w http.ResponseWriter, r *http.Request) {
```
Comment (rutaka-n) [inline user/handler.go]: we need to store the user id of the person who booked the time slot.
e.g. Alice booked time in Bob's calendar.

File: user/handler.go [hunk 5]
```diff
@@ -14,64 +15,65 @@ func Register(w http.ResponseWriter, r *http.Request) {
...
-	}
-
-	if userExists == true {
+	if userExists {
 		w.WriteHeader(http.StatusConflict)
 		return
 	}
 
-	err := repo.Create(u)
+	if userExistsError == sql.ErrNoRows {
```
Comment (rutaka-n) [inline user/handler.go]: `errors.IsError(err, anotherErr)`

File: user/handler.go [hunk 6]
```diff
@@ -14,64 +15,65 @@ func Register(w http.ResponseWriter, r *http.Request) {
...
 	u.Id = uuid.NewString()
 	password, errorSetPassword := u.setPassword()
 	u.Password = password
 
 	if errorSetPassword != nil {
 		w.WriteHeader(http.StatusInternalServerError)
+		return
 	}
 
 	userExists, userExistsError := repo.UserExists(u.Username)
```
Comment (rutaka-n) [inline user/handler.go:27]: if err != nil && errors.Is(err, NoRows) {
    // 500
    return
}
if userExists {
    // Conflict
    return
}

File: user/user.go [hunk 7]
```diff
@@ -5,14 +5,16 @@ import (
...
-	Password  string `json:"password"`
-	Timeslots string `json:"timeslots"`
+	Id       string `json:"id"`
+	Username string `json:"username"`
+	Password string `json:"password"`
 }
 
-type TimeslotStatus struct {
-	Booked bool `json:"booked"`
+type Timeslot struct {
```
Comment (rutaka-n) [inline user/user.go]: maybe move it to separate file/package?

File: user/user.go [hunk 8]
```diff
@@ -5,14 +5,16 @@ import (
...
+	CreatorId     string `json:"CreatorId"`
```
Comment (rutaka-n) [inline user/user.go]: `userID/ownerID/belongsTo` - a user that has the timeslot
`BookedBy` - a user who booked the timeslot.

File: user/handler.go [hunk 9]
```diff
@@ -14,64 +15,65 @@ func Register(w http.ResponseWriter, r *http.Request) {
...
+		err := repo.Create(u)
+
+		if err != nil {
+			w.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+	}
+
+	defaultTimeslot := Timeslot{CreatorId: u.Id, InvitedUserId: "", Time: "", Booked: false}
+	createTimeslotError := repo.CreateTimeslots(defaultTimeslot)
```
Comment (rutaka-n) [inline user/handler.go]: maybe make it in different API endpoint?

File: user/repository.go [hunk 10]
```diff
@@ -16,10 +17,47 @@ func NewRepository(db *sql.DB) *Repository {
...
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (r Repository) isTimeslotBooked(timeslot Timeslot) (Timeslot, error) {
+	ts := Timeslot{}
+	query := "SELECT * FROM timeslots WHERE time=$1 AND invitedUserId=$2"
```
Comment (rutaka-n) [inline user/repository.go]: you can scan into boolean variable.

File: user/user.go (hunk 8 above)
Comment (rutaka-n) [reply user/user.go]: should we add an ID for the timeslot struct?

File: user/handler.go [hunk 11]
```diff
@@ -14,64 +15,65 @@ func Register(w http.ResponseWriter, r *http.Request) {
...
 
-	// user, err := repo.GetById(userId)
+	timeslot, err := repo.isTimeslotBooked(data)
 
-	// if err != nil {
-	// 	log.Panic(err)
-	// }
+	// if there are no rows => timeslots were not created yet at all
+	if err == sql.ErrNoRows {
+		repo.CreateTimeslots(data)
```
Comment (rutaka-n) [inline user/handler.go]: timeslot might be created by the user. and then another user can book it.

File: user/handler.go
Comment (rutaka-n) [inline user/handler.go:1]: Flow:
User1 -> Register()
User1 -> Login()
User1 -> CreateTimeslots()
User2 -> Register()
User2 -> Login()
User2 -> GetUserByName()/SearchUser()/ListUsers()
User2 -> GetTimeslots(User1.ID, from, to) // see list of available timeslots of User1
User2 -> BookTimeSlot(TimeslotID/userID and time)
User1 -> GetTimeslots(User1.ID, from, to) // see list of their timeslots

File: timeslots/handler.go [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+	UserId string `json:"userId"`
+}
+
+func CreateTimeslots(w http.ResponseWriter, r *http.Request) {
+	var timeslotsData TimeslotData
+	json.NewDecoder(r.Body).Decode(&timeslotsData)
+
+	repo := NewRepository(db.DbInstance)
+
+	time := [6]string{"10:00", "11:00", "12:00", "13:00", "14:00", "15:00"}
```
Comment (KiraBurova) [inline timeslots/handler.go:19]: todo:
potentially find something better than just having these types of strings for available timeslots.

Comment (rutaka-n) [reply timeslots/handler.go:19]:
```
parameter: [{from, to}, {from, to}, ...]
```

File: timeslots/handler.go [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+	}
+}
+
+func BookTimeslot(w http.ResponseWriter, r *http.Request) {
+	repo := NewRepository(db.DbInstance)
+
+	var data Timeslot
+	json.NewDecoder(r.Body).Decode(&data)
+
+	ts, err := repo.getTimeslot(data)
```
Comment (rutaka-n) [inline timeslots/handler.go:33]: check if timeslot is absent

File: user/repository.go [hunk 14]
```diff
@@ -58,15 +58,21 @@ func (r Repository) GetByUsername(username string) (User, error) {
 	return user, nil
 }
 
-func (r Repository) UserExists(username string) (bool, error) {
-	user, err := r.GetByUsername(username)
+func (r Repository) userExists(username string) bool {
```
Comment (rutaka-n) [inline user/repository.go:61]: better return error in case database failure.

**Review by rutaka-n — APPROVED** [review]

**Pull request merged**